pkg/submitter/submitter.go

```diff
@@ -0,0 +1,573 @@
+				"solution", result.Nonce,
+				"difficulty", result.Work.Challenge.Difficulty,
+				"requestIDs", fmt.Sprintf("%+v", result.Work.Challenge.RequestIDs),
+			)
+			s.blockUntilTimeToSubmit(ctx)
+			s.handleSubmit(ctx, result)
+		}
+	}
+}
+
+func (s *Submitter) CancelPendingSubmit() {
+	if s.lastSubmitCncl != nil {
+		s.lastSubmitCncl()
+	}
+}
+
+func (s *Submitter) Stop() {
+	s.close()
+}
+
+func (s *Submitter) blockUntilTimeToSubmit(newChallengeReplace context.Context) {
+	var (
+		lastSubmit time.Duration
+		timestamp  *time.Time
+		err        error
+	)
+	for {
+		select {
+		case <-newChallengeReplace.Done():
+		default:
+		}
+		lastSubmit, timestamp, err = s.lastSubmit()
+		if err != nil {
+			level.Debug(s.logger).Log("msg", "checking last submit time", "err", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		break
+	}
+	if lastSubmit < s.cfg.Mine.MinSubmitPeriod.Duration {
+		level.Info(s.logger).Log("msg", "min transaction submit threshold hasn't passed", "minSubmitPeriod", s.cfg.Mine.MinSubmitPeriod, "lastSubmit", lastSubmit)
+		timeToSubmit, cncl := context.WithDeadline(newChallengeReplace, timestamp.Add(15*time.Minute))
+		defer cncl()
+		select {
+		case <-newChallengeReplace.Done():
+		case <-timeToSubmit.Done(): // 15min since last submit has passed to can unblock.
+		}
+	}
+}
+
+func (s *Submitter) handleSubmit(newChallengeReplace context.Context, result *mining.Result) {
+	go func(newChallengeReplace context.Context, result *mining.Result) {
+		ticker := time.NewTicker(1 * time.Second)
+		for {
+			select {
+			case <-newChallengeReplace.Done():
+				level.Info(s.logger).Log("msg", "canceled submit")
+				return
+			default:
+				profitPercent, err := s.profit() // Call it regardless of whether we use so that is sets the exposed metrics.
+				if err != nil {
+					level.Error(s.logger).Log("msg", "submit solution profit check", "err", err)
+					<-ticker.C
+					continue
+				}
+				if s.cfg.Mine.ProfitThreshold > 0 {
+					if profitPercent < int64(s.cfg.Mine.ProfitThreshold) {
+						<-ticker.C
+						continue
+					} else { // Transaction is profitable.
+						for {
+							select {
+							case <-newChallengeReplace.Done():
+								level.Info(s.logger).Log("msg", "canceled submit")
+								return
+							default:
+							}
+							if statusID := s.getMinerStatus(); statusID != 1 {
+								level.Error(s.logger).Log("msg", "miner is not in a status that can submit", "status", minerStatusName(statusID))
+								<-ticker.C
+								continue
+							}
+							s.blockUntilTimeToSubmit(newChallengeReplace)
+							tx, err := s.Submit(newChallengeReplace, result)
+							if err != nil {
+								s.submitFailCount.Inc()
+								level.Error(s.logger).Log("msg", "submiting a solution, retrying", "err", err)
+								<-ticker.C
+								continue
+							}
+							level.Debug(s.logger).Log("msg", "submited a solution", "txHash", tx.Hash().String())
+							s.saveGasUsed(newChallengeReplace, tx)
```

Comment:
Here we want to main context as we don't want to cancel saving the tx cost to the db even when a new challenge arrives. We want to cancel only the submit.
```suggestion
							s.saveGasUsed(s.ctx, tx)
```